sstables_loader.cc

```diff
@@ -255,9 +255,23 @@ future<> tablet_sstable_streamer::stream(bool primary_replica_only) {
             co_await coroutine::maybe_yield();
         }
 
+        llog.debug("Check with tablet_id={} tablet_range={} full_ssts={} part_ssts={}",
+                tablet_id, tablet_range, sstables_fully_contained.size(), sstables_partially_contained.size());
+
         auto tablet_pr = dht::to_partition_range(tablet_range);
-        co_await stream_sstables(tablet_pr, std::move(sstables_partially_contained), primary_replica_only);
-        co_await stream_fully_contained_sstables(tablet_pr, std::move(sstables_fully_contained), primary_replica_only);
+        if (!sstables_partially_contained.empty()) {
+            const auto full_partition_range = dht::partition_range::make_open_ended_both_sides();
+            co_await stream_sstables(full_partition_range, std::move(sstables_partially_contained), primary_replica_only);
+        }
```

Comment:
That's done on purpose. 

we're streaming on a per tablet basis. so when streaming a given tablet, we only consume the range of partially contained sstable that belongs to the tablet in question. In another words, each stream for a tablet X will extract the data from sstables that belongs to X's range.

if we change to full_partition_range, then we're streaming the same sstable multiple times, one for each tablet overlapping with its range. Additionally, it would break load-and-stream in another way for tablets, where a sstable, created in a streaming session, will contain data from different tablets.

See above that we only stop consuming from a sstable once it's considered exhausted, i.e. its last key is before the range of the current tablet being streamed.

Reply:
I am not saying that there's not a bug, just that it's somewhere else in the implementation. How to reproduce it?

Reply:
sstable_it++ is called in two places. 

```
       while (sstable_it != _sstables.rend() && exhausted(*sstable_it)) {
            sstable_it++;
        }

        for (; sstable_it != _sstables.rend(); sstable_it++) {
```

We are done with the sstable as soon as the second loop advances. 


Here is an example:

sst ranges

500 500
400 600
320 350
50 80


tablet ranges
-inf 300
300 500
500 700
700 +inf

The first sst 50 80 will be fully consumed.
we move to second sst 320 350.
The table range will be moved to 300 500
The sst 320 350 will be partially consumed.
We move to the third sst 400 600.
The sst 400 600 will be partially consumed.
We move to the fourth sst  500 550.

Reply:
What is the point of streaming the sstables in multiple stages (each stage streams only the range intersected with the tablet token ranges) if we know the sstable is not fully within a tablet token range.

Reply:
> I am not saying that there's not a bug, just that it's somewhere else in the implementation. How to reproduce it?

1) There is a dtest failure: see https://github.com/scylladb/scylladb/issues/17733

2) Manually

- Create a cluster with tablet enalbed

- Create the tablet table

```
CREATE  ks1.user_table (
    key text, 
    ":attrs" map<text, blob>,
    PRIMARY KEY (key)
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL','rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0
    AND speculative_retry = '99.0PERCENTILE';
```
- Copy the sstable
cp -a scylla-dtest/alternator/snapshot/scylla4.0.rc1/* tmp/1/data/ks1/user_table-e4434dd0e10b11eebc1bd005e06311c0/upload/

- nodetool refresh

- There are 100 partitions in the sstable. See in the log the load and stream stream only part of the partitions due to the bug.

Reply:
> What is the point of streaming the sstables in multiple stages (each stage streams only the range intersected with the tablet token ranges) if we know the sstable is not fully within a tablet token range.

That's because the streaming is per tablet, and when streaming for tablet X, we will only extract from the sstable the range that intersect with tablet range. If we streaming the entire sstable, the destination will receive a sstable that overlaps with more than 1 tablet and that's the main thing load-and-stream for tablets wants to avoid.

I think the bug is in the second loop. we must not advance sstable_it in the second loop, but rather only iterate through a iterator copy. Only the first loop, that checks if sstable is exhaust, must advance sstable_it iterator.

Reply:
This is the patch that fixes it:

```
diff --git a/sstables_loader.cc b/sstables_loader.cc
index d14e3357f8..c43470c5c1 100644
--- a/sstables_loader.cc
+++ b/sstables_loader.cc
@@ -239,8 +239,8 @@ future<> tablet_sstable_streamer::stream(bool primary_replica_only) {
             sstable_it++;
         }
 
-        for (; sstable_it != _sstables.rend(); sstable_it++) {
-            auto sst_token_range = sstable_token_range(*sstable_it);
+        for (auto sst_it = sstable_it; sst_it != _sstables.rend(); sst_it++) {
+            auto sst_token_range = sstable_token_range(*sst_it);
             // sstables are sorted by first key, so we're done with current tablet when
             // the next sstable doesn't overlap with its owned token range.
             if (!tablet_range.overlaps(sst_token_range, dht::token_comparator{})) {
@@ -248,9 +248,9 @@ future<> tablet_sstable_streamer::stream(bool primary_replica_only) {
             }
 
             if (tablet_range.contains(sst_token_range, dht::token_comparator{})) {
-                sstables_fully_contained.push_back(*sstable_it);
+                sstables_fully_contained.push_back(*sst_it);
             } else {
-                sstables_partially_contained.push_back(*sstable_it);
+                sstables_partially_contained.push_back(*sst_it);
             }
             co_await coroutine::maybe_yield();
         }
```

Reply:
@asias Please review: https://github.com/scylladb/scylladb/pull/17899. Was able to write a reproducer in test_tablets.py for the issue.

Reply:
> That's because the streaming is per tablet, and when streaming for tablet X, we will only extract from the sstable the range that intersect with tablet range. If we streaming the entire sstable, the destination will receive a sstable that overlaps with more than 1 tablet and that's the main thing load-and-stream for tablets wants to avoid.

What will happen if we stream ranges covering more than one tablet range on the receiver side? Will the sstable writer split the data automatically?  I think we need to handle this case anyway. For example, after streaming, a tablet range is split. 

I will some more comments regarding  the way we stream. I will continue on your new PR.